lib/libshare/libshare.c

```diff
@@ -78,477 +95,111 @@ register_fstype(const char *name, const sa_share_ops_t *ops)
-		}
-
-		impl_share = next;
-	}
-
-	update_sharetab(impl_handle);
-
-	if (impl_handle->zfs_libhandle != NULL)
-		libzfs_fini(impl_handle->zfs_libhandle);
-
-	impl_share = impl_handle->shares;
-	while (impl_share != NULL) {
-		next = impl_share->next;
-		free_share(impl_share);
-		impl_share = next;
-	}
-
-	free(impl_handle);
-}
-
-static sa_share_impl_t
-find_share(sa_handle_impl_t impl_handle, const char *sharepath)
-{
-	sa_share_impl_t impl_share;
-
-	impl_share = impl_handle->shares;
-	while (impl_share != NULL) {
-		if (strcmp(impl_share->sharepath, sharepath) == 0) {
-			break;
-		}
-
-		impl_share = impl_share->next;
-	}
-
-	return (impl_share);
-}
-
-sa_share_t
-sa_find_share(sa_handle_t handle, char *sharepath)
-{
-	return ((sa_share_t)find_share((sa_handle_impl_t)handle, sharepath));
+	return (found_protocol ? ret : SA_INVALID_PROTOCOL);
 }
 
 int
-sa_enable_share(sa_share_t share, char *protocol)
+sa_disable_share(const char *mountpoint, char *protocol)
 {
-	sa_share_impl_t impl_share = (sa_share_impl_t)share;
 	int rc, ret = SA_OK;
 	boolean_t found_protocol = B_FALSE;
 	sa_fstype_t *fstype;
 
+	sa_share_impl_t impl_share = alloc_share(NULL, mountpoint);
+	if (impl_share == NULL)
+		return (SA_NO_MEMORY);
+
+	(void) libshare_lock();
 	fstype = fstypes;
 	while (fstype != NULL) {
-		if (protocol == NULL || strcmp(fstype->name, protocol) == 0) {
-			update_zfs_share(impl_share, fstype->name);
-
-			rc = fstype->ops->enable_share(impl_share);
+		if (strcmp(fstype->name, protocol) == 0) {
 
+			rc = fstype->ops->disable_share(impl_share);
 			if (rc != SA_OK)
 				ret = rc;
-			else
-				FSINFO(impl_share, fstype)->active = B_TRUE;
 
 			found_protocol = B_TRUE;
 		}
 
 		fstype = fstype->next;
 	}
-
-	update_sharetab(impl_share->handle);
+	(void) libshare_unlock();
+	free_share(impl_share);
 
 	return (found_protocol ? ret : SA_INVALID_PROTOCOL);
 }
 
-int
-sa_disable_share(sa_share_t share, char *protocol)
+boolean_t
+sa_is_shared(const char *mountpoint, char *protocol)
 {
-	sa_share_impl_t impl_share = (sa_share_impl_t)share;
-	int rc, ret = SA_OK;
-	boolean_t found_protocol = B_FALSE;
 	sa_fstype_t *fstype;
+	boolean_t ret = B_FALSE;
+
+	/* guid value is not used */
+	sa_share_impl_t impl_share = alloc_share(NULL, mountpoint);
+	if (impl_share == NULL)
+		return (B_FALSE);
 
 	fstype = fstypes;
 	while (fstype != NULL) {
-		if (protocol == NULL || strcmp(fstype->name, protocol) == 0) {
-			rc = fstype->ops->disable_share(impl_share);
-
-			if (rc == SA_OK) {
-				fstype->ops->clear_shareopts(impl_share);
-
-				FSINFO(impl_share, fstype)->active = B_FALSE;
-			} else
-				ret = rc;
-
-			found_protocol = B_TRUE;
+		if (strcmp(fstype->name, protocol) == 0) {
+			ret = fstype->ops->is_shared(impl_share);
 		}
-
 		fstype = fstype->next;
 	}
+	free_share(impl_share);
+	return (ret);
+}
 
-	update_sharetab(impl_share->handle);
-
-	return (found_protocol ? ret : SA_INVALID_PROTOCOL);
+void
+sa_share_commit(void)
```

Comment:
The other functions seem to follow the naming convention `sa_VERB_OBJECT()`, so how about:
```suggestion
sa_commit_shares(void)
```